scripts/nbd-firewall-config.sh

```diff
@@ -0,0 +1,91 @@
+#!/bin/bash
+#
+# Copyright (c) Citrix Systems. All rights reserved.
+
+##############################################################
+# Use this script to open/close the specified interfaces
+# (IP addresses) for incoming TCP connections to the NBD port.
+#
+# Usage:
+#   ./nbd-firewall-config set [address...]
+#
+##############################################################
+
+# Historical note: the original version of this file filtered
+# by local IP addresses rather than by network interfaces;
+# that version might one day be a useful reference for something.
+
+# Require the "set" argument because otherwise there would be a risk
+# of someone running the script with no arguments and accidentally
+# setting NBD to be available on no addresses.
+# This way, omitting "set" (or supplying "--help") causes the script
+# to print usage instructions.
+
+set -e
+OP="$1"
+if [ _"${OP}" != _set ]; then
+    echo "Usage: $(basename ${0}) set [interface...]
+will set the firewall to allow incoming TCP connections
+to the NBD port on only the specified interfaces (if any)." 1>&2
+    exit 1
+fi
+shift 1
+
+set -eu
+# No automatic cleanup on error because it would depend too
+# critically on the stage at which the error occurs.
+
+NBDPORT=10809
+# Note that chain names must be under 29 characters.
+NEWCHAININ=xapi_nbd_input_chain_new
+NBDCHAININ=xapi_nbd_input_chain
+NEWCHAINOUT=xapi_nbd_output_chain_new
+NBDCHAINOUT=xapi_nbd_output_chain
+
+# Rule fragments for sending NBD packets to NBD chains
+# from INPUT and OUTPUT chains.
+NBD_INPUT_JUMP_TO="-p tcp --dport $NBDPORT -j "
+NBD_OUTPUT_JUMP_TO="-p tcp --sport $NBDPORT -j "
+
+# Same principle as double-buffering a display:
+# add the items one by one to temporary chains,
+# then swap them into use atomically, or as near as possible.
+iptables --new-chain $NEWCHAININ 2>/dev/null || iptables --flush $NEWCHAININ
+iptables --new-chain $NEWCHAINOUT 2>/dev/null || iptables --flush $NEWCHAINOUT
+while [ "$#" -ne 0 ]; do
+    intf="${1}"
+    # (Each positional param is forgotten for ever at the end of the loop.)
+    shift 1
+    iptables -A $NEWCHAININ -m conntrack --ctstate NEW,ESTABLISHED --in-interface "${intf}" -j ACCEPT
+    iptables -A $NEWCHAINOUT --out-interface "${intf}" -j RETURN
+done
+iptables --append $NEWCHAININ -j REJECT
+iptables --append $NEWCHAINOUT -j REJECT
+
+function insert_rule_if_absent {
+    chain="${1}"
+    shift 1
+    rule="${*}"
+    if ! iptables --check "${chain}" $rule 2>/dev/null
+    then
+        iptables --insert "${chain}" $rule
+    fi
+}
+
+# Start using the new chains.
+insert_rule_if_absent INPUT "${NBD_INPUT_JUMP_TO} ${NEWCHAININ}"
+insert_rule_if_absent OUTPUT "${NBD_OUTPUT_JUMP_TO} ${NEWCHAINOUT}"
+
+# If old chains are present, stop using them and then delete them.
+iptables --delete INPUT $NBD_INPUT_JUMP_TO $NBDCHAININ 2>/dev/null || true
+iptables --flush "${NBDCHAININ}" 2>/dev/null && \
```

Comment:
iptables --flush is going to remove all the rules in NBDCHAININ, is it required to a --delete in the previous step ?

Reply:
The aim here is to delete `NBDCHAININ` if it exists.
Before we can do that, we must, in either order:
* remove all the rules in it (hence the `--flush`)
* remove all references to it (hence the `--delete` to remove the referring rule from the `INPUT` chain, if present)

(Note that the `--delete` removes a rule from the `INPUT` chain, not the `NBDCHAININ` chain.)

Reply:
I miss read what "iptables --delete" is trying to delete. lgtm.

Reply:
I've switched from all-caps to camel-case for the chain variables, in hope that it makes the various identifiers a bit more distinct.